### PR TITLE
feat(fslib): recursive rmdir

### PR DESCRIPTION
**Pull request by Larry1123**
**What's the problem this PR addresses?**
@arcanis asked in discord if anyone wanted to add support for recursive since https://github.com/nodejs/node/commit/35b17d9abd123853a8e035ca7bc185f7e942d73b

**How did you fix it?**
`@yarnpkg/fslib`
- `NodeFS` passes opts to node, but only if opts is passed in, otherwise node will fail before 12.10.
- `ZipFS` calls `this.removeSync(p)` if recursive is set to true the options for retry are ignored as it would do nothing for a zip.
- `RmdirOptions` was added to `FakeFS` and an opts was added to `rmdirPromise` and `rmdirSync`
- `ProxifedFS` passes opts as needed to support it
- `ZipOpenFS` passes opts as needed to support it
- `@yarnpkg/fslib` export `RmdirOptions`

`@yarnpkg/pnpfiy`
- `NodeModulesFS` passes opts as needed to support it

**Checklist**
- [x] I have read the [Contributing Guide](https://yarnpkg.com/advanced/contributing).
- [x] I have set the packages that need to be released for my changes to be effective.
- [x] I will check that all automated PR checks pass before the PR gets reviewed.

### CODE REVIEW

File: packages/yarnpkg-fslib/sources/FakeFS.ts [hunk 1]
```diff
@@ -26,6 +26,12 @@ export type MkdirOptions = Partial<{
   mode: number,
 }>;
 
+export type RmdirOptions = Partial<{
```
Comment (paul-soporan) [inline packages/yarnpkg-fslib/sources/FakeFS.ts:29]: `RmdirOptions` should be re-exported inside `index.ts`.

File: packages/yarnpkg-fslib/sources/NodeFS.ts [hunk 2]
```diff
@@ -284,14 +284,20 @@ export class NodeFS extends BasePortableFakeFS {
...
+  async rmdirPromise(p: PortablePath, opts?: RmdirOptions) {
     return await new Promise<void>((resolve, reject) => {
-      this.realFs.rmdir(npath.fromPortablePath(p), this.makeCallback(resolve, reject));
+      // TODO: always pass opts when min node version is 12.10+
+      if (opts) {
+        this.realFs.rmdir(npath.fromPortablePath(p), opts, this.makeCallback(resolve, reject));
+      }
+      else {
+        this.realFs.rmdir(npath.fromPortablePath(p), this.makeCallback(resolve, reject));
+      }
```
Comment (paul-soporan) [inline packages/yarnpkg-fslib/sources/NodeFS.ts:295]: Can you `1tbt`-ify this? (That's the brace style we use in the rest of the codebase, would probably be a good idea to enforce it with an ESLint rule)

**Review by paul-soporan — COMMENTED** [review]
Looks good so far!

Can you also update `NodeModulesFS` from `@yarnpkg/pnpify`? (https://github.com/Larry1123/berry/blob/feat/fslib/recurisive-rmdir/packages/yarnpkg-pnpify/sources/NodeModulesFS.ts#L360)

File: packages/yarnpkg-fslib/sources/NodeFS.ts (hunk 2 above)
Comment (Larry1123) [reply packages/yarnpkg-fslib/sources/NodeFS.ts:295]: You mean put the else on the same line?

Comment (paul-soporan) [reply packages/yarnpkg-fslib/sources/NodeFS.ts]: Yep, the else should be on the same line as the closing if brace.

Comment (Larry1123) [reply packages/yarnpkg-fslib/sources/NodeFS.ts]: ok cool oddly the `1tbt` turned up no results when I searched it.

Comment (paul-soporan) [reply packages/yarnpkg-fslib/sources/NodeFS.ts]: `1tbt` is an acronym for `one true brace style` - https://eslint.org/docs/rules/brace-style

**Comment by arcanis** [comment]
Awesome, thanks! 😃

**Pull request merged**